File: pe-core/src/main/java/org/saar/core/model/ElementsMesh.java
package org.saar.core.model;

import org.saar.lwjgl.opengl.constants.DataType;
import org.saar.lwjgl.opengl.constants.RenderMode;
import org.saar.lwjgl.opengl.objects.vaos.IVao;
import org.saar.lwjgl.opengl.utils.GlRendering;

public class ElementsMesh extends MeshBase implements Mesh {

    private final RenderMode mode;
    private final int count;
    private final DataType indexType;
    private final long indices;

    public ElementsMesh(IVao vao, RenderMode renderMode, int count, DataType indexType) {
        super(vao);
        this.mode = renderMode;
        this.count = count;
        this.indexType = indexType;
        this.indices = 0;
    }

    @Override
    protected void doDrawCall() {
        GlRendering.drawElements(this.mode, this.count, this.indexType, this.indices);
    }
}

File: pe-core/src/main/java/org/saar/core/mesh/ElementsMesh.kt
package org.saar.core.mesh

import org.saar.lwjgl.opengl.constants.DataType
import org.saar.lwjgl.opengl.constants.RenderMode
import org.saar.lwjgl.opengl.drawcall.ElementsDrawCall
import org.saar.lwjgl.opengl.vao.IVao

class ElementsMesh(
    private val vao: IVao,
    var drawCall: ElementsDrawCall
) : Mesh {

    fun set(
        renderMode: RenderMode = this.drawCall.renderMode,
        count: Int = this.drawCall.count,
        type: DataType = this.drawCall.type,
        indices: Long = this.drawCall.indices,
    ) {
        this.drawCall = ElementsDrawCall(renderMode, count, type, indices)
    }

    override fun draw() {
        this.vao.bind()
        this.vao.enableAttributes()
        this.drawCall.doDrawCall()
    }

    override fun delete() {
        this.vao.delete()
    }
}